wireapp/wire-android
# feat: migrate cryptobox and client data [AR-2591] [AR-2658]

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-2591" title="AR-2591" target="_blank"><img alt="Subtask" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10816?size=medium" />AR-2591</a>  migrate users crypto box
  </td></table>
  <br />
 








----
#### PR Submission Checklist for internal contributors

- The **PR Title**

- The **PR Description**
----

# What's new in this PR?

### Issues

In order to successfully perform a migration from scala app, we need to copy cryptobox files to the appropriate destination and registered client data to the new database.|


### Solutions

Added `isMigrationCompletedFlow` to `ConnectionPolicyManager` not to `startObservingAppLifecycle` until the migration is completed.
Added `ScalaCryptoBoxDirectoryProvider` to get the path of the cryptobox files.
Updated `ScalaUserDatabaseProvider` to be able to get database for multiple users.
Refactored `MigrateActiveAccountsUseCase` to return list of migrated user ids.
Implemented `MigrateClientsDataUseCase` which moves cryptobox files to the proper destination, adds registered client ids and creates the user session to perform a sync after that.

### Testing

#### How to Test

Install and log in to the scala app and update to the AR version, the app should perform the migration and be already logged in and able to send and receive messages.

----
#### PR Post Submission Checklist for internal contributors (Optional)

----
#### PR Post Merge Checklist for internal contributors

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.
